pycolmap/pycolmap/feature/sift.h

```diff
@@ -0,0 +1,118 @@
+#include "colmap/feature/sift.h"
+#include "colmap/feature/utils.h"
+
+#include "pycolmap/helpers.h"
+#include "pycolmap/utils.h"
+
+#include <memory>
+
+#include <Eigen/Core>
+#include <FreeImage.h>
+#include <pybind11/numpy.h>
+#include <pybind11/pybind11.h>
+
+#define kdim 4
+
+using namespace colmap;
+using namespace pybind11::literals;
+namespace py = pybind11;
+
+template <typename dtype>
+using pyimage_t =
+    Eigen::Matrix<dtype, Eigen::Dynamic, Eigen::Dynamic, Eigen::RowMajor>;
+
+typedef Eigen::Matrix<float, Eigen::Dynamic, Eigen::Dynamic, Eigen::RowMajor>
+    descriptors_t;
+typedef Eigen::Matrix<float, Eigen::Dynamic, kdim, Eigen::RowMajor> keypoints_t;
+typedef std::tuple<keypoints_t, descriptors_t> sift_output_t;
+
+static std::map<int, std::unique_ptr<std::mutex>> sift_gpu_mutexes;
+
+class Sift {
+ public:
+  Sift(SiftExtractionOptions options, Device device)
+      : options_(std::move(options)), use_gpu_(IsGPU(device)) {
+    VerifyGPUParams(use_gpu_);
+    options_.use_gpu = use_gpu_;
+    extractor_ = CreateSiftFeatureExtractor(options_);
+    THROW_CHECK(extractor_ != nullptr);
+  }
+
+  sift_output_t Extract(const Eigen::Ref<const pyimage_t<uint8_t>>& image) {
+    THROW_CHECK_LE(image.rows(), options_.max_image_size);
+    THROW_CHECK_LE(image.cols(), options_.max_image_size);
+
+    const unsigned int bpp = 8;  // Grey.
+    const unsigned int width = image.cols();
+    const unsigned int scan_width = (bpp / 8) * width;
+    pyimage_t<uint8_t> image_copy = image;
+    FIBITMAP* bitmap_raw = FreeImage_ConvertFromRawBitsEx(
+        /*copySource=*/false,
+        static_cast<unsigned char*>(image_copy.data()),
+        FIT_BITMAP,
+        width,
+        image.rows(),
+        scan_width,
+        bpp,
+        FI_RGBA_RED_MASK,
+        FI_RGBA_GREEN_MASK,
+        FI_RGBA_BLUE_MASK,
+        /*topdown=*/true);
+    const Bitmap bitmap(bitmap_raw);
+
+    FeatureKeypoints keypoints_;
+    FeatureDescriptors descriptors_;
+    THROW_CHECK(extractor_->Extract(bitmap, &keypoints_, &descriptors_))
+    const size_t num_features = keypoints_.size();
+
+    keypoints_t keypoints(num_features, kdim);
+    for (size_t i = 0; i < num_features; ++i) {
+      keypoints(i, 0) = keypoints_[i].x;
+      keypoints(i, 1) = keypoints_[i].y;
+      keypoints(i, 2) = keypoints_[i].ComputeScale();
+      keypoints(i, 3) = keypoints_[i].ComputeOrientation();
+    }
+
+    descriptors_t descriptors = descriptors_.cast<float>();
+    descriptors /= 512.0f;
+
+    return std::make_tuple(keypoints, descriptors);
```

Comment:
return `std::make_tuple(std::move(keypoints), std::move(descriptors))` will avoid copying the data.

Reply:
Good catch, done.